interface/ceed.c

```diff
@@ -751,6 +788,81 @@ int CeedReference(Ceed ceed) {
   return CEED_ERROR_SUCCESS;
 }
 
+/**
+  @brief Get a `CeedVector` for scratch work from a `Ceed` context.
+
+  Note: This vector must be restored with @ref CeedRestoreWorkVector().
+
+  @param[in]  ceed `Ceed` context
+  @param[in]  len  Minimum length of work vector
+  @param[out] vec  Address of the variable where `CeedVector` will be stored
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Backend
+**/
+int CeedGetWorkVector(Ceed ceed, CeedSize len, CeedVector *vec) {
+  CeedInt i = 0;
+
+  if (!ceed->work_vectors) CeedCall(CeedWorkVectorsCreate(ceed));
+
+  // Search for big enough work vector
+  for (i = 0; i < ceed->work_vectors->num_vecs; i++) {
+    if (!ceed->work_vectors->is_in_use[i]) {
+      CeedSize work_len;
+
+      CeedCall(CeedVectorGetLength(ceed->work_vectors->vecs[i], &work_len));
+      if (work_len >= len) break;
+    }
+  }
+  // Long enough vector was not found
+  if (i == ceed->work_vectors->num_vecs) {
+    if (ceed->work_vectors->max_num_vecs == 0) {
+      ceed->work_vectors->max_num_vecs = 1;
+      CeedCall(CeedCalloc(ceed->work_vectors->max_num_vecs, &ceed->work_vectors->vecs));
+      CeedCall(CeedCalloc(ceed->work_vectors->max_num_vecs, &ceed->work_vectors->is_in_use));
+    } else if (ceed->work_vectors->max_num_vecs == i) {
+      ceed->work_vectors->max_num_vecs *= 2;
+      CeedCall(CeedRealloc(ceed->work_vectors->max_num_vecs, &ceed->work_vectors->vecs));
+      CeedCall(CeedRealloc(ceed->work_vectors->max_num_vecs, &ceed->work_vectors->is_in_use));
+    }
+    ceed->work_vectors->num_vecs++;
+    CeedCallBackend(CeedVectorCreate(ceed, len, &ceed->work_vectors->vecs[i]));
+    ceed->ref_count--;  // Note: ref_count manipulation to prevent a ref-loop
+  }
+  // Return pointer to work vector
+  ceed->work_vectors->is_in_use[i] = true;
+  *vec                             = NULL;
```

Comment:
May be misremembering, but don't we normally call `CeedVectorDestroy()` rather than setting to `NULL`?

Reply:
This is due to a C-ism.

The variable may be uninitialized and contain junk, but if we call RreferenceCopy on junk then we'll segfault. We can either set the pointer to NULL ourselves before ReferenceCopy (current code), or require the user to pass in a valid pointer to a vec they haven't used yet, or we could use Reference with a manual copy.

Reply:
I've thought about of making a helper for the combo of
```C
foo = NULL;
Ceed*ReferenceCopy(bar, &foo);
```
but the main reason I haven't is because I really don't want users generally doing such a thing as that becomes a possible leak

Reply:
Ah, yeah, I [ran into that for `ReferenceCopy`](https://gitlab.com/phypid/honee/-/blob/main/src/qdata.c?ref_type=heads#L61-64). It's not a big ask for users to initialize those as `NULL`.